pkg/api/helpers_test.go

```diff
@@ -397,6 +398,275 @@ func TestMatchTaint(t *testing.T) {
+			},
+			expectTolerated: false,
+		},
+		{
+			description: "toleration with explicit forgiveness can't tolerate taint with no added time, expect not tolerated",
+			toleration: Toleration{
+				Key:                "foo",
+				Operator:           TolerationOpExists,
+				Effect:             TaintEffectNoExecute,
+				ForgivenessSeconds: genForgivenessSeconds(300),
+			},
+			taint: Taint{
+				Key:    "foo",
+				Effect: TaintEffectNoExecute,
+			},
+			expectTolerated: false,
+		},
+	}
+	for _, tc := range testCases {
+		if tolerated := tc.toleration.ToleratesTaint(&tc.taint); tc.expectTolerated != tolerated {
+			t.Errorf("[%s] expect %v, got %v: toleration %+v, taint %s", tc.description, tc.expectTolerated, tolerated, tc.toleration, tc.taint.ToString())
+		}
+	}
+}
+
+func TestTolerationsTolerateTaintsWithFilter(t *testing.T) {
+	testCases := []struct {
+		description       string
+		tolerations       []Toleration
+		taints            []Taint
+		isInterestedTaint taintsFilterFunc
+		expectTolerated   bool
+	}{
+		{
+			description:       "empty tolerations tolerate emtpy taints",
+			tolerations:       []Toleration{},
+			taints:            []Taint{},
+			isInterestedTaint: func(t Taint) bool { return true },
+			expectTolerated:   true,
+		},
+		{
+			description: "non-empty tolerations tolerate emtpy taints",
+			tolerations: []Toleration{
+				{
+					Key:      "foo",
+					Operator: "Exists",
+					Effect:   TaintEffectNoSchedule,
+				},
+			},
+			taints:            []Taint{},
+			isInterestedTaint: func(t Taint) bool { return true },
+			expectTolerated:   true,
+		},
+		{
+			description: "tolerations match all taints, expect tolerated",
+			tolerations: []Toleration{
+				{
+					Key:      "foo",
+					Operator: "Exists",
+					Effect:   TaintEffectNoSchedule,
+				},
+			},
+			taints: []Taint{
+				{
+					Key:    "foo",
+					Effect: TaintEffectNoSchedule,
+				},
+			},
+			isInterestedTaint: func(t Taint) bool { return true },
+			expectTolerated:   true,
+		},
+		{
+			description: "tolerations don't match taints, but no taint is interested, expect tolerated",
+			tolerations: []Toleration{
+				{
+					Key:      "foo",
+					Operator: "Exists",
+					Effect:   TaintEffectNoSchedule,
+				},
+			},
+			taints: []Taint{
+				{
+					Key:    "bar",
+					Effect: TaintEffectNoSchedule,
+				},
+			},
+			isInterestedTaint: func(t Taint) bool { return false },
+			expectTolerated:   true,
+		},
+		{
+			description: "no isInterestedTaint indicated, means all taints are interested, tolerations don't match taints, expect untolerated",
+			tolerations: []Toleration{
+				{
+					Key:      "foo",
+					Operator: "Exists",
+					Effect:   TaintEffectNoSchedule,
+				},
+			},
+			taints: []Taint{
+				{
+					Key:    "bar",
+					Effect: TaintEffectNoSchedule,
+				},
+			},
+			isInterestedTaint: nil,
+			expectTolerated:   false,
+		},
+		{
+			description: "tolerations match interested taints, expect tolerated",
```

Comment:
@ingvagabund I don't understand the point you're making. Can you be more specific about what scenario you're concerned about? BTW note that `isInterestedTaint` is not part of the API for taints -- it is part of the internal implementation.